### PR TITLE
Add CITATION file (and copy it in) (Closes: #3423)

### PR DESCRIPTION
**Pull request by eddelbuettel**
This PR implements the proposed addition of a CITATION file.   From R it allows us to generate citation and text and bibtex entries easily:

```r
> citation("mlpack")

To cite package ‘mlpack’ in publications use:

  Singh Parihar Y, Curtin R, Eddelbuettel D, Balamuta J (2023). _mlpack: 'Rcpp' Integration for the 'mlpack' Library_. R package version 4.0.1,
  <https://CRAN.R-project.org/package=mlpack>.

  Curtin R, Edel M, Shrit O, Agrawal S, Basak S, Balamuta J, Birmingham R, Dutt K, Eddelbuettel D, Garg R, Jaiswal S, Kaushik A, Kim S, Mukherjee A, Sai N, Sharma N, Parihar Y,
  Swain R, Sanderson C (2023). “mlpack 4: a fast, header-only C++ machine learning library.” _Journal of Open Source Software_, *8*(82). doi:10.21105/joss.05026
  <https://doi.org/10.21105/joss.05026>.

To see these entries in BibTeX format, use 'print(<citation>, bibtex=TRUE)', 'toBibtex(.)', or set 'options(citation.bibtex.max=999)'.

> print(citation("mlpack"), bibtex=TRUE)

To cite package ‘mlpack’ in publications use:

  Singh Parihar Y, Curtin R, Eddelbuettel D, Balamuta J (2023). _mlpack: 'Rcpp' Integration for the 'mlpack' Library_. R package version 4.0.1,
  <https://CRAN.R-project.org/package=mlpack>.

A BibTeX entry for LaTeX users is

  @Manual{,
    title = {mlpack: 'Rcpp' Integration for the 'mlpack' Library},
    author = {Yashwant {Singh Parihar} and Ryan Curtin and Dirk Eddelbuettel and James Balamuta},
    year = {2023},
    note = {R package version 4.0.1},
    url = {https://CRAN.R-project.org/package=mlpack},
  }

  Curtin R, Edel M, Shrit O, Agrawal S, Basak S, Balamuta J, Birmingham R, Dutt K, Eddelbuettel D, Garg R, Jaiswal S, Kaushik A, Kim S, Mukherjee A, Sai N, Sharma N, Parihar Y,
  Swain R, Sanderson C (2023). “mlpack 4: a fast, header-only C++ machine learning library.” _Journal of Open Source Software_, *8*(82). doi:10.21105/joss.05026
  <https://doi.org/10.21105/joss.05026>.

A BibTeX entry for LaTeX users is

  @Article{,
    title = {mlpack 4: a fast, header-only C++ machine learning library},
    author = {Ryan R. Curtin and Marcus Edel and Omar Shrit and Shubham Agrawal and Suryoday Basak and James J. Balamuta and Ryan Birmingham and Kartik Dutt and Dirk Eddelbuettel and Rishabh Garg and Shikhar Jaiswal and Aakash Kaushik and Sangyeon Kim and Anjishnu Mukherjee and Nanubala Gnana Sai and Nippun Sharma and Yashwant Singh Parihar and Roshan Swain and Conrad Sanderson},
    journal = {Journal of Open Source Software},
    year = {2023},
    volume = {8},
    number = {82},
    doi = {10.21105/joss.05026},
  }

> 
```

Thanks for the hint in https://github.com/mlpack/mlpack/issues/3423#issuecomment-1439167508 :grinning: I also added the `cmake` copy step.   Sorry for the brief delay in shipping this but hey vacation beckoned for once...

### CODE REVIEW

File: src/mlpack/bindings/R/mlpack/inst/CITATION [hunk 1]
```diff
@@ -0,0 +1,49 @@
+bibentry("Manual",
+         other = unlist(citation(auto = meta), recursive = FALSE))
```
Comment (coatless) [inline src/mlpack/bindings/R/mlpack/inst/CITATION:2]: Nice trick! I've not seen that one before.

Looks like the delayed promise execution causes the `meta` object to be automatically defined when `citation()` is called? 

```r
if (!is.null(auto) && !is.logical(auto) && !anyNA(match(c("Package", 
        "Version", "Title"), names(meta <- as.list(auto)))) && 
        !all(is.na(match(c("Authors@R", "Author"), names(meta))))) {
        auto_was_meta <- TRUE
        package <- meta$Package
    }
```

**Review by coatless — APPROVED** [review]
Looks great to me!

File: src/mlpack/bindings/R/mlpack/inst/CITATION (hunk 1 above)
Comment (eddelbuettel) [reply src/mlpack/bindings/R/mlpack/inst/CITATION:2]: I _literally_ stole^Hborrowed this whole first entry for 'the package' from one of Kurt's own packages and its CITATION file, site unseen. Worked when I tried it and carried it over to two or three my packages.

How to automagically get the package name within the scope of a file in the package had bugged me for some time.   I still do it 'explicitly' in the testrunner invocation in `tests/*R` which is generally tinytest for me.

**Comment by eddelbuettel** [comment]
> Looks great to me!

As I / we had the `paper.md` file from the paper repo, it was really just a matter of a quick Emacs keystroke macro to reshape each entry accordingly for its `person()` representation.

**Comment by rcurtin** [comment]
Thanks!  Hope the vacation was good. :)

**Review by rcurtin — APPROVED** [review]

**Pull request merged**